kmod/src/dir.c

```diff
@@ -449,67 +450,114 @@ static int scoutfs_readdir(struct file *file, struct dir_context *ctx)
 	struct scoutfs_dirent *dent = NULL;
 	struct scoutfs_key last_key;
 	struct scoutfs_key key;
-	int name_len;
-	u64 pos;
+	struct page *page = NULL;
+	void *data = NULL;
+	int *name_len;
+	u64 page_pos;
 	int ret;
+	int complete = 0;
+	int i, j;
 
 	if (!dir_emit_dots(file, ctx))
 		return 0;
 
-	dent = alloc_dirent(SCOUTFS_NAME_LEN);
-	if (!dent) {
+	page = alloc_page(GFP_KERNEL);
+	if (!page)
 		return -ENOMEM;
-	}
+	data = page_address(page);
 
 	init_dirent_key(&last_key, SCOUTFS_READDIR_TYPE, scoutfs_ino(inode),
 			SCOUTFS_DIRENT_LAST_POS, 0);
 
+	/*
+	 * lock and fetch dirent items, until the page no longer fits
+	 * a max size dirent (288b). Then unlock and dir_emit the ones
+	 * we stored in the page.
+	 *
+	 * We store the name_len first (4b) and then the dirent. page_pos
+	 * is maintained to keep progress. Once done fetching keys, we mark
+	 * complete. The second iteration will stop once it reachs the count
+	 * from the first iteration.
+	 *
+	 * If the dirents are ~20b size, we should get about 50-100 in a
+	 * single page.
+	 */
+cont:
 	ret = scoutfs_lock_inode(sb, SCOUTFS_LOCK_READ, 0, inode, &dir_lock);
 	if (ret)
-		goto out;
+		goto out_page;
 
-	for (;;) {
+	page_pos = 0;
+
+	for (i = 0; page_pos + sizeof(int) + dirent_bytes(SCOUTFS_NAME_LEN) < PAGE_SIZE; i++) {
 		init_dirent_key(&key, SCOUTFS_READDIR_TYPE, scoutfs_ino(inode),
 				ctx->pos, 0);
 
+		dent = (struct scoutfs_dirent *)(data + page_pos + sizeof(int));
```

Comment:
The main problem with all this is unaligned accesses.  dirent_bytes() can be byte aligned, so the name_len and dent pointers can end up unaligned.  That's fine on x86, but very not fine on everything else.  (And may break on debug tests depending on how aggressively unforgiving they are.)  I think the fix here is a lot like fiemap.  An outer loop with two inner loops to get rid of the gotos.  And you can just iterate through dent pointers if you have a dirent_bytes() variant that aligns.  Have an end that's page_address(page) + PAGE_SIZE.  Then it can fill while next_aligned_dirent(dent, SCOUTFS_NAME_LEN) < end.  After reading each you can have dent = next_aligned_dirent(dent, len).  And you can store name_len in one of the dirent _pad fields to avoid storing it separately.  Finally, it needs to be more careful about setting ctx->pos.  The caller will unconditionally set the file->pos to ctx->pos, so it should really only be set after successfully emiting an entry.  The setting of it to the next entry about to be returned is a micro optimization that we could just leave out if it's too confusing.  And we don't need all the out_ exit labels.  Set the lock to NULL after you unlock it in the null and an _unlock in the exit path will be safe along with kfree() and the guarded page free so everything can always use the same exit path.

Reply:
same here, we'll redo the review.